docs/plutus-ir/plutus-ir.md

```diff
@@ -0,0 +1,193 @@
+            )
+        )
+    )
+)
+```
+
+
+## Validity
+
+### Mutually recursive types
+
+The syntax of Plutus IR suggests that you can write mutually recursive datatypes, even though
+we don't know how to compile those yet. I think we will eventually be able to cover all
+types defined with this syntax, but until then, validity of Plutus IR programs will be somewhat
+determined simply by *whatever we are able to compile*.
+
+### Typechecking
+
+Similarly, while we will define a typing judgement for Plutus IR, we will probably not
+write a typechecker for it, and rather will compile it to Plutus Core before typechecking.
+
+### Value restriction
+
+Since type-lets are going to be desugared as type abstractions, a type-let whose body is not
+a value will fail the Plutus Core value restriction, and there is no attempt to prevent this
+in the Plutus IR syntax. It will simply compile into an invalid Plutus Core program.
+
+## Typing
+
+Typing rules are largely the same as Plutus Core.
+
+(I should write out some real rules for this, but everything being variadic makes it a bit noisy.)
+
+Let bindings are straightforward: they introduce names into the context with their declared
+type or kind. For recursive let bindings we typecheck each binding under the assumption that
+all the bindings have their declared type or kind.
+
+We do need to state something more complex about datatype bindings: they introduce a
+number of names with various types or kinds, with the correctness conditions being:
+- The type variable declarations are well-kinded.
+- The application of the datatype to its type variables is well-kinded and of kind `(type)`.
+- The constructor argument types are all well-kinded and of type `(type)`.
+
+## Design notes
+
+### Why do we need variadic syntax?
+
+`letrec` is inherently about binding an arbitrary number of mutually recursive bindings, so
+we need to be able to write that.
+
+### Why is `let` variadic?
+
+`letrec` obviously needs to be variadic, since the whole point is to have multiple bindings
+that reference each other.
+
+`let` doesn't *need* to be variadic, but it's symmetrical, convenient, and no harder to support.
+
+### Why is there a special datatype binding?
+
+Datatype bindings are weird in that they bind multiple names, which are part of the binding
+(the type, the constructors, the destructor). So we can't handle it as a normal "type binding" since
+it also declares other things.
+
+### Why are variable declarations reified in the syntax?
+
+It's very helpful to have a syntax form for variable declarations, since the syntax is too
+homogeneous when everything is just names. For example, in the datatype declaration we have
+a whole series of type variable declarations. I think we could just about parse the version
+with them all flattened, but it's much clearer when they're reified.
+
+I think this is quite useful and arguably worth introducing in the Plutus Core syntax too.
```

Comment:
Yes, I guess that's true. It's mostly consistency, and I like the additional structure. It's nice to have binders be obvious because they're the things with variable declarations in them.